token-erc20/chaincode-javascript/lib/tokenERC20.js

```diff
@@ -0,0 +1,367 @@
+        }
+
+        // Update the balance
+        const fromUpdatedBalance = fromCurrentBalance - valueInt;
+        const toUpdatedBalance = toCurrentBalance + valueInt;
+
+        await ctx.stub.putState(fromBalanceKey, Buffer.from(fromUpdatedBalance.toString()));
+        await ctx.stub.putState(toBalanceKey, Buffer.from(toUpdatedBalance.toString()));
+
+        console.log(`client ${from} balance updated from ${fromCurrentBalance} to ${fromUpdatedBalance}`);
+        console.log(`recipient ${to} balance updated from ${toCurrentBalance} to ${toUpdatedBalance}`);
+
+        return true;
+    }
+
+    /**
+     * Allows `spender` to spend `value` amount of tokens from the owner.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} spender The spender
+     * @param {Integer} value The amount of tokens to be approved for transfer
+     * @returns {Boolean} Return whether the approval was successful or not
+     */
+    async approve(ctx, spender, value) {
+        const owner = ctx.clientIdentity.getID();
+
+        const allowanceKey = ctx.stub.createCompositeKey(allowancePrefix, [owner, spender]);
+        if (!allowanceKey) {
+            throw new Error('Failed to create the composite key');
+        }
+
+        let valueInt = parseInt(value);
+        await ctx.stub.putState(allowanceKey, Buffer.from(valueInt.toString()));
+
+        // Emit the Approval event
+        const approvalEvent = { owner, spender, value: valueInt };
+        ctx.stub.setEvent('Approval', Buffer.from(JSON.stringify(approvalEvent)));
+
+        console.log('approve ended successfully');
+        return true;
+    }
+
+    /**
+     * Returns the amount of tokens which `spender` is allowed to withdraw from `owner`.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} owner The owner of tokens
+     * @param {String} spender The spender who are able to transfer the tokens
+     * @returns {Number} Return the amount of remaining tokens allowed to spent
+     */
+    async allowance(ctx, owner, spender) {
+        const allowanceKey = ctx.stub.createCompositeKey(allowancePrefix, [owner, spender]);
+        if (!allowanceKey) {
+            throw new Error('Failed to create the composite key');
+        }
+
+        const allowanceBytes = await ctx.stub.getState(allowanceKey);
+        const allowance = parseInt(allowanceBytes.toString());
+        return allowance;
+    }
+
+    // ================== Extended Functions ==========================
+
+    /**
+     * Set optional infomation for a token.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} name The name of the token
+     * @param {String} symbol The symbol of the token
+     * @param {String} decimals The decimals of the token
+     * @param {String} totalSupply The totalSupply of the token
+     */
+    async setOption(ctx, name, symbol, decimals) {
```

Comment:
This is just my idea because ERC-20 spec does not explain how to set options. If there is a better approach, I would like to change it.
OpenZeppelin's ERC20 implementation sets name, symbol and decimals in the constructor of the Solidity contract. For Chaincode, does this mean to set options in the Init() function?